src/control-plane/backend/lambda/api/service/segments/user-segments-sql.ts

```diff
@@ -0,0 +1,462 @@
+      }
+    }).join(' JOIN ');
+    const whereClause = events.map((event, index) => (`e${index + 1}.event_name = '${event.eventName}'`)).join(' AND ');
+
+    return `
+        SELECT DISTINCT e1.user_pseudo_id
+        FROM ${fromClause}
+        WHERE ${whereClause}
+    `;
+  }
+
+  /**
+   * Build UserAttributeCondition sql query statement
+   * @param condition
+   * @private
+   */
+  private buildUserAttributeConditionStatement(condition: UserAttributeCondition) {
+    const table = this.segment.appId + '.' + USER_TABLE;
+    const userHasAttribute = `
+        SELECT DISTINCT user_pseudo_id
+        FROM ${table}
+        WHERE ${this.composeParameterConditionClause(condition.attributeCondition, 'UserProperty')}
+    `;
+
+    return condition.hasAttribute ? userHasAttribute : `
+        SELECT DISTINCT u.user_pseudo_id
+        FROM ${table} u
+                 LEFT JOIN (${userHasAttribute}) uha ON u.user_pseudo_id = uha.user_pseudo_id
+        WHERE uha.user_pseudo_id IS NULL
+    `;
+  }
+
+  /**
+   * Build UserInSegmentCondition sql query statement
+   * @param condition
+   * @private
+   */
+  private buildUserInSegmentConditionStatement(condition: UserInSegmentCondition) {
+    const userTable = this.segment.appId + '.' + USER_TABLE;
+    const segmentTable = this.segment.appId + '.' + SEGMENT_USER_TABLE;
+    return `
+        SELECT DISTINCT user_pseudo_id
+        FROM ${userTable}
+        WHERE user_pseudo_id ${condition.isInSegment ? '' : 'NOT'} IN (
+            SELECT user_id
+            FROM ${segmentTable}
+            WHERE segment_id = '${condition.segmentId}'
+        )
+    `;
+  }
+
+  /**
+   * Compose date range clause from SegmentFilterGroup. Note that date range setting is optional.
+   * @param group SegmentFilterGroup
+   * @private
+   */
+  private composeDateRangeClause(group: SegmentFilterGroup) {
+    const { isRelativeDateRange, startDate, endDate, lastN, timeUnit } = group;
+
+    // Date range is optional
+    if ((isRelativeDateRange && (lastN === undefined || timeUnit === undefined))
+      || (!isRelativeDateRange && (startDate === undefined || endDate === undefined))) {
+      return 'event_timestamp >= GETDATE() - INTERVAL \'7 day\''; // last 7 days by default
+    }
+
+    if (isRelativeDateRange) {
+      return `event_timestamp >= GETDATE() - INTERVAL '${lastN} ${timeUnit}'`;
+    } else {
+      return `event_timestamp BETWEEN '${startDate}' AND '${endDate}'`;
+    }
+  }
+
+  /**
+   * Compose number of days clause from SegmentFilterGroup.
+   * @param group SegmentFilterGroup
+   * @private
+   */
+  private composeNumberOfDaysClause(group: SegmentFilterGroup) {
+    const { isRelativeDateRange, startDate, endDate, lastN, timeUnit } = group;
+
+    if (isRelativeDateRange && lastN !== undefined && timeUnit !== undefined) {
+      return `DATEDIFF(DAY, GETDATE() - INTERVAL '${lastN} ${timeUnit}', GETDATE())`;
```

Comment:
should use the timezone of the app.

Reply:
The relative time range (lastN timeUnit) supports minute, hour, day, month, year etc. I think it's either confusing to apply timezone to *day* only, or hard to apply to all the time units. Maybe we can support timezone in frontend UI as user select absolute time range.